suffuse/opts.go

```diff
@@ -2,86 +2,138 @@ package suffuse
+  fs.StringVar(&opts.VolName, "n", "", "volume name (OSX only)")
+  fs.BoolVar(&opts.Scratch, "t", false, "create scratch directory as mount point")
+  fs.BoolVar(&opts.Verbose, "v", false, "log at INFO level")
+  fs.BoolVar(&opts.Debug, "d", false, "log at DEBUG level")
+
+  err := fs.Parse(args[1:])
+  if err != nil {
+    // Parse has already printed the problem and usage, there is no way to tell it not to do that
+    return nil, &SfsConfigError{err: err.Error(), usage: nil}
   }
 
-  c, err := fuse.Mount(mnt.Path, mount_opts...)
-  if err != nil { return Sfs{}, err }
+  opts.Args = fs.Args()
+  return opts, nil
+}
+
+func validate(opts *sfsOpts) *SfsConfigError {
+  exists := func (path string, msg string) *SfsConfigError {
+    if (!NewPath(path).FileExists()) { return opts.newError(Sprintf(msg, path)) }
+    return nil
+  }
 
   // Exactly one incoming path for the moment.
   // Eventually more than one could mean a union mount.
-  if len(opts.Args) != 1 { UsageAndExit() }
+  if len(opts.Args) != 1 { return opts.newError("Suffuse currently accepts a single path") }
 
-  mfs := Sfs {
-    Mountpoint : mnt,
-    RootNode   : NewIdNode(NewPath(opts.Args[0])),
-    Connection : c,
-  }
+  if err := exists(opts.Mountpoint, "Mountpoint '%s' does not exist" ); err != nil { return err }
+  if err := exists(opts.Config    , "Config file '%s' does not exist"); err != nil && opts.Config != "" { return err }
 
-  /** Start a goroutine which looks for INT/TERM and
-   *  calls unmount on the filesystem. Otherwise ctrl-C
-   *  leaves us with ghost mounts which outlive the process.
-   */
-  trap := func (sig os.Signal) {
-    Echoerr("Caught %s - forcing unmount(2) of %s\n", sig, mfs.Mountpoint)
-    mfs.Unmount()
+  for _, path := range opts.Args {
+    if err := exists(path, "Path '%s' does not exist"); err != nil { return err }
   }
-  TrapExit(trap)
-  return mfs, nil
+
+  return nil
 }
 
-func (opts SfsOpts) GetMountpoint() Path {
-  switch {
-    case opts.Scratch          : return ScratchDir()
-    case opts.Mountpoint != "" : return NewPath(opts.Mountpoint)
-    default                    : UsageAndExit() ; return Path{}
+func configFromOpts(opts *sfsOpts) (*SfsConfig, *SfsConfigError) {
+
+  determineMountPoint := func() (Path, *SfsConfigError) {
+    switch {
+      case opts.Scratch          : return ScratchDir(), nil
+      case opts.Mountpoint != "" : return NewPath(opts.Mountpoint), nil
+      default                    : return Path{}, opts.newError("Mountpoint can not be empty if the scratch flag is false")
+    }
+  }
+
+  determineLogLevel := func() lg.Lvl {
+    switch {
+      case opts.Debug   : return lg.LvlDebug
+      case opts.Verbose : return lg.LvlInfo
+      default           : return lg.LvlWarn
+    }
+  }
+
+  determineConfig := func() *Path {
+    if opts.Config == "" {
+      return nil
+    } else {
+      var c = NewPath(opts.Config)
+      return &c
+    }
   }
```

Comment:
These seem like gratuitous closures, especially as much boilerplate as go requires. I understand the desire to write little helper functions but the penalty is much higher in go than scala and I don't see either of these carrying their weight. Contrast with:
```go
  var determineConfig *Path
  if opts.Config != "" {
    path := NewPath(opts.Config)
    determineConfig = &path
  }
```
And it's only that long because I had to do the stupid assign-and-then-take-address. We should probably use *Path everywhere. If not then a NewPathRef which returns *Path.